lms/djangoapps/courseware/masquerade.py

```diff
@@ -40,6 +40,16 @@ def __init__(self, course_key, role='student', user_partition_id=None, group_id=
         self.group_id = group_id
         self.user_name = user_name
 
+    def __setstate__(self, state):
+        """
+        Ensure that all attributes are initialised when unpickling CourseMasquerade objects.
+
+        Users might still have CourseMasquerade objects from older versions of the code in their
+        session.  These old objects might not have all attributes set, possibly resulting in
+        AttributeErrors.
+        """
+        self.__init__(**state)
```

Comment:
@chrisndodge Since we haven't defined `__getstate__()`, the default implementation will be used, which simply returns the instance's `__dict__`, so `state` should never contain any fields that are not allowed as parameters to `__init__()`.

The idea of this fix is to (1) mitigate an issue some users were currently experiencing with old `CourseMasquerade` objects unpickled from their session and (2) prevent similar issues from happening in the future, if someone adds further fields to `CourseMasquerade`.  What you are pointing out could become an issue in the future if someone *removes* a field from `CourseMasquerade` without updating this method.  We could protect against that contingency by adding a dummy `**kwargs` to `__init__()`, but I don't think we should.  If anything, we should add a comment somewhere explaining what to do to remove an attribute.